### DISCUSSION TITLE
Web3Modal V3 - Attempting eager MetaMask connection before user triggers modal open.

### DISCUSSION THREAD

**Discussion by saintmaxi** [WalletConnect/walletconnect-monorepo#3937] [Developer Support]
I am developing a web app using Web3Modal V3 with the SvelteKit framework and am seeing an issue where MetaMask is being automatically prompted for a connection before the user interacts with the modal to actually trigger it there. I find this to be negative for user experience and am struggling to figure out why this is happening. It appears that it specifically happens when I disconnect in MetaMask rather than the modal, then reload the page. If I disconnect in the modal first then this does not happen. Is this expected? Testing out another example I have using V2 does not seem to cause the issue.

Here is a POC I put together: https://rad-rugelach-aff2ed.netlify.app/ ([Source Code](https://github.com/saintmaxi/web3modal-eager-connection-poc))

**Steps to Reproduce:**
1. Click button.
2. Connect with injected MetaMask via Web3Modal.
3. Disconnect from site using MetaMask UI (not Disconnect button in modal).
4. Reload page, observe eager connection.

I'm assuming that because the modal session is not disconnected, it attempts the autoConnect logic configured in the modal resulting in the MetaMask prompt. However, I thought that my logic in watchAccount would circumvent this.

**EDIT:** I am trying to use `modal.disconnect()` directly rather than the wagmi `disconnect` in my example based on the docs [here](https://docs.walletconnect.com/web3modal/javascript/actions#disconnect), but I am getting the error that web3modal.disconnect() is not a function. Printing out all the properties/functions available on the web3modal object shows that disconnect is not an available method. Is this a bug?

```js
import { createWeb3Modal, defaultWagmiConfig } from '@web3modal/wagmi'
import { watchAccount, watchNetwork, disconnect, switchNetwork, getPublicClient, getWalletClient } from '@wagmi/core'
import { getContract } from 'viem';
import { mainnet } from 'viem/chains'
import { get, writable } from "svelte/store";

// Wallet connection status (true/false)
export const connected = writable(undefined);
// Current wallet address
export const walletAddress = writable(undefined);
// Current public client
export const publicClient = writable(undefined);
// Current wallet client 
export const walletClient = writable(undefined);
// Current network ID
export const networkID = writable(undefined);

// 1. Define constants
const projectId = 'INSERT_ID_HERE'

// 2. Create wagmiConfig
const metadata = {
  name: 'Web3Modal',
  description: 'Web3Modal Example',
  url: 'https://web3modal.com',
  icons: ['https://avatars.githubusercontent.com/u/37784886']
}

const chains = [mainnet]
const wagmiConfig = defaultWagmiConfig({ chains, projectId, metadata })

// 3. Create modal
const web3modal = createWeb3Modal({ wagmiConfig, projectId, chains })

watchAccount(async (account) => {
    if (account?.isConnected) {
      connected.set(account.isConnected);
      const _publicClient = getPublicClient({chainId: mainnet.id})
      publicClient.set(_publicClient);
      const _walletClient = await getWalletClient({chainId: mainnet.id});
      walletClient.set(_walletClient);
      walletAddress.set(account.address);
    }
    else {
      connected.set(false);
      web3modal.disconnect();
    }
});
  
watchNetwork(async (network) => {
    networkID.set(network?.chain?.id);
    const _publicClient = getPublicClient({chainId: mainnet.id})
    publicClient.set(_publicClient);
    const _walletClient = await getWalletClient({chainId: mainnet.id});
    walletClient.set(_walletClient);
});
  
export const handleConnectWallet = async () => {
    web3modal.open();
};
```

**Comment by glitch-txs** [comment]
thanks for the feedback and steps to repro! I'm sending this to Ilja and we should fix it soon

**Reply by glitch-txs** [reply]
you can keep track of the issue [here](https://github.com/WalletConnect/web3modal/issues/1488)
I sent it to the Web3Modal maintainer

**Reply by saintmaxi** [reply]
thanks all! i appreciate it

**Comment by ethan-auroria** [comment]
@saintmaxi if you haven't found the solution to the missing `disconnect` function, I have found the workaround:

```javascript
	function onAccountChanged(account: GetAccountResult) {
		if (account.address !== undefined || account.isConnected) {
			account.connector?.disconnect();

			goto('/');
		}

		// I'm not sure what should be done here, because we are assuming that there is an account connected.
	}

	onMount(() => {
		const metadata = {...};
		const chains = [...];
		const wagmiConfig = defaultWagmiConfig({ chains, projectId: PUBLIC_WALLETCONNECT_PROJECT_ID, metadata });

		createWeb3Modal({ wagmiConfig, projectId: PUBLIC_WALLETCONNECT_PROJECT_ID, chains });
		unwatchAccount = watchAccount(onAccountChanged);

		return () => {
			// Remove callbacks.
		};
	});
```
